R/convert_output_types.R

```diff
@@ -0,0 +1,251 @@
+      }
+    model_out_tbl_transform[[i]] <- convert_from_sample(
+      grouped_model_out_tbl, new_output_type[i], new_output_type_id_tmp
+    )
+  }
+  return(dplyr::bind_rows(model_out_tbl_transform))
+}
+
+validate_new_output_type <- function(starting_output_type, new_output_type,
+                                     new_output_type_id) {
+  valid_conversions <- list(
+    "sample" = c("mean", "median", "quantile", "cdf"),
+    "quantile" = c("mean", "median", "cdf"),
+    "cdf" = c("mean", "median", "quantile")
+  )
+  # check starting_output_type is supported
+  valid_starting_output_type <- starting_output_type %in% names(valid_conversions)
+  if (!valid_starting_output_type) {
+    cli::cli_abort(c(
+      "{.var output_type} provided cannot be transformed",
+      i = "must be of type {.var sample}, {.var quantile}, {.var cdf}."
+    ))
+  }
+  # check new_output_type is supported
+  invalid_new_output_type <- which(!(new_output_type %in% valid_conversions[[starting_output_type]]))
+  if (length(invalid_new_output_type) > 0) {
+    cli::cli_abort(c(
+      "{invalid_new_output_type} cannot be transformed to the specified
+      {.var new_output_type}",
+      i = "new_output_type must be {valid_conversions[[starting_output_type]]}"
+    ))
+  }
+  # check new_output_type_id
+  if (all(new_output_type %in% c("mean", "median")) && !all(is.na(new_output_type_id))) {
+    cli::cli_abort(c(
+      "{.var new_output_type_id} is incompatible with {.var new_output_type}",
+      i = "{.var new_output_type_id} should be {.var NA}"
+    ))
+  } else if ("quantile" %in% new_output_type) {
+    new_output_type_id_tmp <- new_output_type_id
+    if (is.list(new_output_type_id)) {
+      new_output_type_id_tmp <- new_output_type_id[["quantile"]]
+    }
+    if (!is.numeric(new_output_type_id_tmp)) {
+      cli::cli_abort(c(
+        "elements of {.var new_output_type_id} should be numeric",
+        i = "elements of {.var new_output_type_id} represent quantiles
+                of the predictive distribution"
+      ))
+    }
+    if (any(new_output_type_id_tmp < 0) || any(new_output_type_id_tmp > 1)) {
+      cli::cli_abort(c(
+        "elements of {.var new_output_type_id} should be between 0 and 1",
+        i = "elements of {.var new_output_type_id} represent quantiles
+                of the predictive distribution"
+      ))
+    }
+  } else if (new_output_type == "cdf") {
+    new_output_type_id_tmp <- new_output_type_id
+    if (is.list(new_output_type_id)) {
+      new_output_type_id_tmp <- new_output_type_id[["cdf"]]
+    }
+    if (!is.numeric(new_output_type_id_tmp)) {
+      cli::cli_abort(c(
+        "elements of {.var new_output_type_id} should be numeric",
+        i = "elements of {.var new_output_type_id} represent possible
+                values of the target variable"
+      ))
+    }
+  }
+}
+
+#' @export
+get_samples_from_quantiles <- function(model_out_tbl, task_id_cols, n_samples, ...) {
+  set.seed(101)
```

Comment:
I don't think we should be setting the seed internally, rather this should be sth a user should be able to control. Is there a reason you have set it?

Also, In any functions using distfromq, as we are just adding it as a Suggests dependency, it's a good idea to add the above check to ensure it's installed and fail gracefully if not.

```suggestion
  if (!requireNamespace("distfromq")) {
    cli::cli_abort(
      c("x" = "{.pkg distfromq} must be installed to convert {.val cdf} 
                   or {.val quantile} output types.",
        "i" = "Use {.code remotes::install_github('reichlab/distfromq')} to install."
      )
    )
  }
```
```